chronos/asyncmacro2.nim

```diff
@@ -48,21 +27,40 @@ proc completeWithNode(fut, baseType, node: NimNode): NimNode {.compileTime.} =
           ident "is", nnkTypeOfExpr.newTree(node), ident "void"),
         newStmtList(
           node,
-          completeWithResult(fut, baseType)
         )
       ),
       nnkElseExpr.newTree(
-        newCall(ident "complete", fut, node)
+        newAssignment(resultNode, node)
       )
     )
 
-proc processBody(node, fut, baseType: NimNode): NimNode {.compileTime.} =
+proc processBody(node, resultNode, baseType: NimNode): NimNode {.compileTime.} =
+  # Rewrites the returns to fill the future instead
   #echo(node.treeRepr)
   case node.kind
   of nnkReturnStmt:
-    let
-      res = newNimNode(nnkStmtList, node)
-    res.add completeWithNode(fut, baseType, processBody(node[0], fut, baseType))
+    let res = newNimNode(nnkStmtList, node)
+    if node[0].kind != nnkEmpty:
+      # transforms return to:
+      # when `baseType` isnot void:
+      #   `resultNode` = processBody(node)
+      # else:
+      #   {.error: "no async return type declared".}
+      # return
+      res.add nnkWhenStmt.newTree(
+        nnkElifExpr.newTree(
+          nnkInfix.newTree(
+            ident "isnot", baseType, ident "void"),
+          newAssignment(resultNode, processBody(node[0], resultNode, baseType))
+        ),
+        nnkElse.newTree(
+          nnkPragma.newTree(
+            nnkExprColonExpr.newTree(
+              newIdentNode("error"),
+              newLit("no async return type declared"))
```

Comment:
This is confusing error message for everybody, even i do not understand what it mean.

Reply:
```
proc x: void =
  return 42
```
```
testit.nim(2, 3) Error: no return type declared
```

this is what the compiler normally prints - not great but... maybe "cannot return value in function returning void"

also, the call to `error` should include `node` so that line number is included

Reply:
And lesson learned, shouldn't copy the compiler error messages :stuck_out_tongue:

Reply:
Compiler provides `no return type declared` your code provides `no async return type declared`.
Where you found any mention of `async return type` and what the difference with `return type` is unclear.

Reply:
A lot of compiler error messages are "cryptic" and sometimes when you generate code using macro, compiler could generate absolutely inadequate error message which could be understand only by macro code author. In this case generating error message is author's duty, because at the end error message will be read by library user, and it should not be cryptic, it should has meaning in context.